ktor-server/ktor-server-plugins/ktor-server-rate-limit/jvmAndNix/src/io/ktor/server/plugins/ratelimit/RateLimit.kt

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright 2014-2022 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.server.plugins.ratelimit
+
+import io.ktor.http.*
+import io.ktor.server.application.*
+import io.ktor.server.response.*
+import io.ktor.server.routing.*
+import io.ktor.util.*
+import io.ktor.util.collections.*
+import io.ktor.util.date.*
+import io.ktor.util.logging.*
+import kotlin.time.*
+
+internal val LOGGER = KtorSimpleLogger("io.ktor.server.plugins.RateLimit")
+private const val LIMITER_NAME_GLOBAL = "KTOR_GLOBAL_RATE_LIMITER"
+internal const val LIMITER_NAME_EMPTY = "KTOR_NO_NAME_RATE_LIMITER"
+
+internal val RateLimiterInstancesRegistryKey =
+    AttributeKey<ConcurrentMap<Pair<String, Any>, RateLimiter>>("RateLimiterInstancesRegistryKey")
+
+internal val RateLimiterConfigsRegistryKey =
+    AttributeKey<Map<String, RateLimitProvider>>("RateLimiterConfigsRegistryKey")
+
+/**
+ * A plugin that provides rate limiting for incoming requests.
+ */
+public val RateLimit: ApplicationPlugin<RateLimitConfig> = createApplicationPlugin("RateLimit", ::RateLimitConfig) {
+    val global = pluginConfig.global
+    val providers = when {
+        global != null -> pluginConfig.providers.toMap() + (LIMITER_NAME_GLOBAL to global)
+        else -> pluginConfig.providers.toMap()
+    }
+    check(providers.isNotEmpty()) { "At least one provider must be specified" }
+    application.attributes.put(RateLimiterConfigsRegistryKey, providers)
+
+    if (global == null) return@createApplicationPlugin
+    application.install(RateLimitApplicationInterceptors) {
+        this.providerNames = listOf(LIMITER_NAME_GLOBAL)
+    }
+}
+
+/**
+ * A config for [RateLimit] plugin.
+ */
+@KtorDsl
+public class RateLimitConfig {
+
+    internal val providers: MutableMap<String, RateLimitProvider> = mutableMapOf()
+    internal var global: RateLimitProvider? = null
+
+    /**
+     * Registers Rate-Limit provider that can be used in sub routes via [Route.rateLimit] function.
+     */
+    public fun register(name: String = LIMITER_NAME_EMPTY, block: RateLimitProviderConfig.() -> Unit) {
+        if (providers.containsKey(name)) {
+            throw IllegalStateException("Rate limit provider with name $name is already configured")
+        }
+        providers[name] = RateLimitProvider(
+            RateLimitProviderConfig().apply {
+                this.name = name
+                block()
+            }
+        )
+    }
+
+    /**
+     * Registers Rate-Limit provider that is applied to whole application
+     */
+    public fun global(block: RateLimitProviderConfig.() -> Unit) {
+        global = RateLimitProvider(
+            RateLimitProviderConfig().apply {
+                this.name = LIMITER_NAME_GLOBAL
+                block()
+            }
+        )
+    }
+}
+
+/**
+ * A config for [RateLimit] provider inside [RateLimiterConfig].
+ */
+@KtorDsl
+public class RateLimitProviderConfig {
```

Comment:
Could you tell me how we explain having provider config public, without explaining what provider is?

Reply:
@andreyaksenov what do you think?

Reply:
Having the [RateLimiterConfig] internal class in a description of a public class might confuse.